linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java

```diff
@@ -1322,9 +1351,10 @@ private void closeInner() {
       final Enumerable<TSource> outer, final Enumerable<TInner> inner,
       final Predicate2<TSource, TInner> predicate,
       Function2<TSource, TInner, TResult> resultSelector,
-      final boolean generateNullsOnLeft,
-      final boolean generateNullsOnRight) {
+      final JoinType joinType) {
```

Comment:
I was thinking that EnumerableDefaults is an important API for Linq4j project. For somebody who uses only this part of Calcite I would say that it is part of the public API so this could be considered a breaking change. Since the other reviewers (@julianhyde @danny0405  @hsyuan ) didn't bring this up I guess I am overreacting so you can ignore my comment.

Reply:
To be on the safe side, we could keep the old method as deprecated, I'm not against it.
However, considering that this method was already renamed as part of [[CALCITE-2969] Improve design of join-like relational expressions](https://issues.apache.org/jira/browse/CALCITE-2969) (previously it was named thetaJoin), I think we have already "broken" many things for 1.20, so I guess one more little change little will not be that relevant.

Reply:
@zabetak One reason I didn't deprecate the redundant link4j CorrelateJoinType/JoinType in CALCITE-2969 is that calcite-link4j is a separate/single module from calcite-core, while the calcite-core depends on calcite-link4j, so we may need to keep the redundant, it's not that bad because these interfaces are not ambiguous and users of other system would not use them actually.

I suggest to not let this be blocker for merging this in 1.20.0

Reply:
I'm with you @danny0405 but this PR introduces another JoinType in Linq4j module.

Reply:
Got your idea, i’m +1 with your proposal.